### PR TITLE
DiffBind - rework to make collection-friendly and other improvements

### PR DESCRIPTION
**Pull request by mblue9**
Inspired by the Great Work of @jmchilton on the new Rule Builder for collections, I found some fresh enthusiasm to go back and retackle diffbind - the most collection-unfriendly of all the Galaxy tools I know. As diffbind's current wrapper only allows collections to have one dataset, which is awkward and very limiting. And there's also an excessive amount of manual typing/copy-paste required for each sample in the tool form.  See issue https://github.com/galaxyproject/tools-iuc/issues/1694

ping @zfrenchee

### CODE REVIEW

**Comment by mblue9** [comment]
Thanks for the review @bgruening! I found some other things that look like they could be changed to improve the wrapper, but at the risk of turning this into another bomb of unrelated commits and delaying things, I've put them into an issue here: https://github.com/galaxyproject/tools-iuc/issues/1809

**Comment by bgruening** [comment]
No hurry from my side :)

**Comment by mblue9** [comment]
Ok cool, if noone objects I might try to add some of the https://github.com/galaxyproject/tools-iuc/issues/1809 changes here then, which I could group loosely into an atomic PR called "diffbind - rework to make collection-friendly and other improvements"  :). As I need to use this tool and would prefer if it had some of those things incorporated. Will aim to have the extra changes in by sometime next week.

**Comment by zfrenchee** [comment]
One small request on my side: If the FDR cutoff could be a text field rather than a slider that could be nice.

**Comment by mblue9** [comment]
@zfrenchee the FDR cutoff is currently both a slider and a text box and you can use either, see here:

screen shot 2018-04-05 at 1 57 17 pm

Is that ok?

File: tools/diffbind/diffbind.xml [hunk 1]
```diff
@@ -143,10 +147,13 @@ Rscript '$__tool_directory__/diffbind.R'
         <data name="rdata" format="rdata" from_work_dir="DiffBind_analysis.RData" label="${tool.name} on ${on_string}: RData file">
             <filter>out['rdata']</filter>
         </data>
+        <data name="rscript" format="txt" from_work_dir="*.txt" label="${tool.name} on ${on_string}: Rscript">
```
Comment (bgruening) [inline tools/diffbind/diffbind.xml]: I guess the `from_work_dir` is not needed here?

File: tools/diffbind/diffbind.R [hunk 2]
```diff
@@ -6,6 +6,7 @@ Sys.setlocale("LC_MESSAGES", "en_US.UTF-8")
 suppressPackageStartupMessages({
 	library('getopt')
 	library('DiffBind')
+    library('rjson')
```
Comment (bgruening) [inline tools/diffbind/diffbind.R]: indentation

File: tools/diffbind/diffbind.xml [hunk 3]
```diff
@@ -223,7 +223,11 @@ Rscript '$__tool_directory__/diffbind.R'
             <output name="binding_matrix" value="out_binding.matrix" />
             <output name="rdata" value="DiffBind_analysis.RData" compare="sim_size"/>
             <output name="rscript" value="out_rscript.txt"/>
-            <output name="analysis_info" value="out_analysis_info.txt"/>
+            <output name="analysis_info" value="out_analysis_info.txt" compare="sim_size" >
```
Comment (bgruening) [inline tools/diffbind/diffbind.xml]: I think if you have assert_content you don't need to have sime_size? Or is both, the sim_size and the assert, checked in this case?

File: tools/diffbind/diffbind.xml [hunk 4]
```diff
@@ -19,69 +20,143 @@
...
+            <option value="narrow" selected="true">narrowPeaks</option>
+            <option value="macs">MACS .xls</option>
+            <option value="swembl">SWEMBL</option>
+            <option value="bayes">bayesPeak</option>
+            <option value="fp4">FindPeaks v4</option>
+         </param>
+         <param name="peakformat" type="select" optional="True" label="Peak File Format" help="Specify format for peak files. Default: narrowPeaks">
+            <option value="raw">raw (scoreCol=4)</option>
+            <option value="bed">bed (scoreCol=5)</option>
+            <option value="narrow" selected="true">narrowPeaks (scoreCol=8)</option>
```
Comment (bgruening) [inline tools/diffbind/diffbind.xml]: Narrowpeaks is also from MACS, right?

File: tools/diffbind/diffbind.xml [hunk 5]
```diff
@@ -19,69 +20,143 @@
...
+                <param name="bamreads" type="data" format="bam" multiple="true" label="Read BAM file" help="Specify the Read BAM file used for Peak calling."/>
+                <param name="bamcontrol" type="data" format="bam" multiple="true" optional="True" label="Control BAM file" help="If specifying a control BAM file, all samples are required to specify one."/>
+            </repeat>
         </repeat>
-        <param name="th" type="float" value="1" min="0" max="1"
-                label="FDR Threshold"
-                help="Significance threshold; all sites with FDR less than or equal to this value will be included in the report. A value of 1 will include all binding sites in the report. Default: 1"/>
-        
+
+        <param name="peakcaller" type="select" optional="True" label="Peak Caller" help="Specify peak caller. If Peaks is not a bed file, this will determine how the Peaks file is parsed. Default: narrowPeaks">
```
Comment (bgruening) [inline tools/diffbind/diffbind.xml]: What is the difference between the format and the caller? Currently I can mix a caller with a different format, right? Let's say bayesPeak with SCIER.

**Comment by bgruening** [comment]
This gets me super excited :)

File: tools/diffbind/diffbind.xml (hunk 4 above)
Comment (mblue9) [reply tools/diffbind/diffbind.xml]: I've removed this peakFormat parameter and the peakCaller one below also as you just confirmed what I was wondering, adding these params is probably only going to create more confusion than they're worth. These options are described in the diffbind vignette but I think just the scorecol is all that's really important. Thanks for confirming :smile:

File: tools/diffbind/diffbind.xml (hunk 5 above)
Comment (mblue9) [reply tools/diffbind/diffbind.xml]: removed, see comment above

**Comment by mblue9** [comment]
Yes I'm also excited about this one, I'd never used diffbind til I started working on this wrapper and now I can see it looks very useful, well worth the wrapping effort :smile:

File: tools/diffbind/diffbind.xml (hunk 3 above)
Comment (bgruening) [reply tools/diffbind/diffbind.xml]: I was just wondering if Galaxy is such a case really checks to tests at once the test for similar size and the assert. If so I was not aware of this :)

**Comment by bgruening** [comment]
Yeah, sorry for the first version :)
Let get this deployed.

File: tools/diffbind/diffbind.xml (hunk 3 above)
Comment (mblue9) [reply tools/diffbind/diffbind.xml]: well I just added it in and received no complaints from planemo, got the green light so went with it..but does it actually work..hmm

**Pull request merged**